managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpReadEntry.java

```diff
@@ -60,14 +60,15 @@ public static OpReadEntry create(ManagedCursorImpl cursor, PositionImpl readPosi
     @Override
     public void readEntriesComplete(List<Entry> returnedEntries, Object ctx) {
         // Filter the returned entries for individual deleted messages
-        final Position nexReadPosition = returnedEntries.get(returnedEntries.size() - 1).getPosition().getNext();
+        final PositionImpl lastPosition = (PositionImpl) returnedEntries.get(returnedEntries.size() - 1).getPosition();
         if (log.isDebugEnabled()) {
             log.debug("[{}][{}] Read entries succeeded batch_size={} cumulative_size={} requested_count={}",
                     cursor.ledger.getName(), cursor.getName(), returnedEntries.size(), entries.size(), count);
         }
         List<Entry> filteredEntries = cursor.filterReadEntries(returnedEntries);
         entries.addAll(filteredEntries);
 
+        final Position nexReadPosition = getNextAvailablePosition(lastPosition);
```

Comment:
We should try to avoid checking the range set if there are no filtered entries. There are some objects allocated in there when checking the ranges (which we should address at some point, though outside the scope of this change).